specification/flag-evaluation/flag-evaluation.md

```diff
@@ -0,0 +1,195 @@
+# Flag Evaluation API
+
+## Overview
+
+The `evaluation API` allows for the retrieval of feature flag values, independent of any flag control plane or vendor. In the absence of a feature flag system implementation (a [provider](../provider/providers.md)) the `evaluation API` uses the "No-op provider", which simply returns the supplied default flag value.
+
+#### Initialization and Configuration
+
+##### Requirement 1.1
+
+> The `API`, and any state it maintains **SHOULD** exist as a global singleton, even in cases wherein multiple versions of the `API` are present at runtime.
+
+It's important that multiple instances of the `API` not be active, so that state stored therein, such as the registered `provider`, static global `evaluation context`, and globally configured `hooks` allow the `API` to behave predictably. This can be difficult in some runtimes or languages, but implementors should make their best effort that only a single instance of the `API` is used.
+
+##### Requirement 1.2
+
+> The `API` **MUST** provide a function to set the `provider`, which accepts an API-conformant `provider` implementation.
+
+```
+// example provider mutator
+OpenFeature.setProvider(new MyProvider());
+```
+
+see [provider](../provider//providers.md)
+
+##### Requirement 1.3
+
+> The `API` **MUST** provide a function for `hook` addition, which accepts an array of API-conformant `hooks`. This adds the specified hooks to any hooks already attached.
```

Comment:
I've intentionally not gone into detail here, since `hooks` will have their own section.

Reply:
Specifying it must be an array is an implementation detail IMO and depends on the idioms of the implementation language. In some languages it is idiomatic to have variadic arguments for example.

```suggestion
> The `API` **MUST** provide a function to register one or more `hooks`. This adds the specified hooks to any hooks already attached.
```

"This adds the specified hooks to any hooks already attached" reads weird to me. I sort of get what you're going for (that it doesn't replace existing hooks, but includes new ones) but it seems like a truism "addHooks adds hooks"

```suggestion
> The `API` **MUST** provide a function for `hook` addition, which accepts an array of API-conformant `hooks`. When new hooks are added, previously added hooks are not removed.
```

This also makes execution order ambiguous. I would either specify that it is not order guaranteed, or use a word like "append" or add a sentence like "hooks are executed in the order they are added".

```suggestion
> The `API` **MUST** provide a function for `hook` addition, which accepts an array of API-conformant `hooks`. This adds the specified hooks to any hooks already attached. Hooks are executed in the order they are added.
```

And all together:

```suggestion
> The `API` **MUST** provide a function to register one or more `hooks` which accepts an array of API-conformant `hooks`. When a new hook is added, previously added hooks are not removed. Hooks are executed in the order they are added.
```

Reply:
@dyladan  Good feedback here, I've amalgamated a few of these and updated.

We will have an entire hook section, and I wanted to talk about ordering in that document in some more detail. To that end, we'd discussed `before` hooks running in the order they are added, with all other hooks (`after`, `error`, `finally`, which run after evaluation) running in reverse order. Do you think that's too much to add here?